File: feature_tests/dart/lib/src/OptionOpaque.dart
import 'OptionStruct.dart';
import 'dart:ffi' as ffi;
import 'diplomat_runtime.dart';

typedef OptionOpaqueFfi = ffi.Pointer<ffi.Opaque>;
class OptionOpaque implements ffi.Finalizable {

  final OptionOpaqueFfi _underlying;

  
static OptionOpaque? new(int i) { 
    final result = _newFfi(i);
    return result.address == 0 ? null : OptionOpaqueFromFfi(result);
  }
  static late final _newFfi = capi<ffi.NativeFunction<OptionOpaqueFfi Function(ffi.Int32)>>('OptionOpaque_new')
  .asFunction<OptionOpaqueFfi Function(int)>();

  
static OptionOpaque? newNone() { 
    final result = _newNoneFfi();
    return result.address == 0 ? null : OptionOpaqueFromFfi(result);
  }
  static late final _newNoneFfi = capi<ffi.NativeFunction<OptionOpaqueFfi Function()>>('OptionOpaque_new_none')
  .asFunction<OptionOpaqueFfi Function()>();

  
static OptionStruct newStruct() { 
    final result = _newStructFfi();
    return OptionStructFromFfi(result);
  }
  static late final _newStructFfi = capi<ffi.NativeFunction<OptionStructFfi Function()>>('OptionOpaque_new_struct')
  .asFunction<OptionStructFfi Function()>();

  
static OptionStruct newStructNones() { 
    final result = _newStructNonesFfi();
    return OptionStructFromFfi(result);
  }
  static late final _newStructNonesFfi = capi<ffi.NativeFunction<OptionStructFfi Function()>>('OptionOpaque_new_struct_nones')
  .asFunction<OptionStructFfi Function()>();

  
void assertInteger(int i) { 
    _assertIntegerFfi(this._underlying,i);
  }
  static late final _assertIntegerFfi = capi<ffi.NativeFunction<ffi.Void Function(OptionOpaqueFfi, ffi.Int32)>>('OptionOpaque_assert_integer')
  .asFunction<void Function(OptionOpaqueFfi, int)>();

  
static bool optionOpaqueArgument(OptionOpaque? arg) { 
    final result = _optionOpaqueArgumentFfi(OptionOpaque?AsFfi(arg));
    return result;
  }
  static late final _optionOpaqueArgumentFfi = capi<ffi.NativeFunction<ffi.Bool Function(OptionOpaqueFfi)>>('OptionOpaque_option_opaque_argument')
  .asFunction<bool Function(OptionOpaqueFfi)>();

  OptionOpaque._(this._underlying) {
    _finalizer.attach(this, this._underlying.cast());
  }

  static late final _finalizer = ffi.NativeFinalizer(capi<ffi.NativeFunction<ffi.Void Function(ffi.Pointer<ffi.Void>)>>('OptionOpaque_destroy'));
}

// These are not methods because we want to keep them package-private, and methods are either private or public
OptionOpaque OptionOpaqueFromFfi(OptionOpaqueFfi underlying) => OptionOpaque._(underlying);
OptionOpaqueFfi OptionOpaqueAsFfi(OptionOpaque t) => t._underlying;

File: feature_tests/js/api/OptionOpaque.mjs
import wasm from "./diplomat-wasm.mjs"
import * as diplomatRuntime from "./diplomat-runtime.mjs"
import { OptionStruct } from "./OptionStruct.mjs"

const OptionOpaque_box_destroy_registry = new FinalizationRegistry(underlying => {
  wasm.OptionOpaque_destroy(underlying);
});

export class OptionOpaque {
  #lifetimeEdges = [];
  constructor(underlying, owned, edges) {
    this.underlying = underlying;
    this.#lifetimeEdges.push(...edges);
    if (owned) {
      OptionOpaque_box_destroy_registry.register(this, underlying);
    }
  }

  static new(arg_i) {
    return (() => {
      const option_ptr = wasm.OptionOpaque_new(arg_i);
      return (option_ptr == 0) ? null : new OptionOpaque(option_ptr, true, []);
    })();
  }

  static new_none() {
    return (() => {
      const option_ptr = wasm.OptionOpaque_new_none();
      return (option_ptr == 0) ? null : new OptionOpaque(option_ptr, true, []);
    })();
  }

  static new_struct() {
    return (() => {
      const diplomat_receive_buffer = wasm.diplomat_alloc(16, 4);
      wasm.OptionOpaque_new_struct(diplomat_receive_buffer);
      const out = new OptionStruct(diplomat_receive_buffer);
      wasm.diplomat_free(diplomat_receive_buffer, 16, 4);
      return out;
    })();
  }

  static new_struct_nones() {
    return (() => {
      const diplomat_receive_buffer = wasm.diplomat_alloc(16, 4);
      wasm.OptionOpaque_new_struct_nones(diplomat_receive_buffer);
      const out = new OptionStruct(diplomat_receive_buffer);
      wasm.diplomat_free(diplomat_receive_buffer, 16, 4);
      return out;
    })();
  }

  assert_integer(arg_i) {
    wasm.OptionOpaque_assert_integer(this.underlying, arg_i);
  }

  static option_opaque_argument(arg_arg) {
    return wasm.OptionOpaque_option_opaque_argument(arg_arg);
  }
}
